### PR TITLE
dine-travel

### PR DESCRIPTION
**Pull request by MarcelGarus**
We opened an easier-to-read issue, #9

From @batteredgherkin and me

image

### CODE REVIEW

File: cp23/hpi-fun2/dinetravel/executables/generator.cpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+                int n = MAX;
+
+                ll timeline[n];
+                rep(n, i) timeline[i] = i;
+                // Fisher-Yates-shuffle, https://stackoverflow.com/questions/22850316/how-to-shuffle-elements-in-a-vector-randomly
+                rep(n, k)
+                {
+                    ll r = k + rnd.next(MAX) % (n - k);
+                    swap(timeline[k], timeline[r]);
+                }
```
Comment (schmidtjonas) [inline cp23/hpi-fun2/dinetravel/executables/generator.cpp]:
```suggestion
                shuffle(timeline.begin(), timeline.end());
```

File: .template/problem.tex [hunk 2]
```diff
@@ -6,17 +6,19 @@
...
+\placeholder{Input description}
+
 \paragraph*{Input}
 
 \placeholder{Input description}
 
 \paragraph*{Output}
 
-\placeholder{Output description}
+\placeholder{Input description}
```
Comment (schmidtjonas) [inline .template/problem.tex]: Please don't change the template

File: cp23/hpi-fun2/dinetravel/problem.tex [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+As everyone knows, time travel is a delicate thing. In order to precisely alter history for her dream dish, Mabel now wants to analyze the closest thing there is to the fruit lasagna in this world, the vegetable lasagna.
+As a thorough researcher, she consults many vegetable lasagna historians and notes all important events in a timeline. The historians also inform her about causal dependencies between those events, meaning some event A caused another event B.
+
+When Mabel evaluates all her notes, she realizes: She’s so close! She has uncovered almost the entire history of the vegetable lasagna’s origin. Only a single causal connection is missing and thereby prevents her from understanding the concept of vegetable lasagna on a more fundamental, holistic level.
+
+To estimate how much research still needs to be done, she wants to find the number of possibilities for the last connection so that it still conforms to the timeline given by the historians as well as the other known connections. She knows all the events that she noted directly or indirectly caused vegetable lasagna to emerge.
```
Comment (schmidtjonas) [inline cp23/hpi-fun2/dinetravel/problem.tex:16]: I don't understand completely. So the inserted edge has to conform to the topological order. Is the last sentence supposed to say that every node should have a path to the lasagna? Can you clarify that in the statement?

Comment (MarcelGarus) [reply cp23/hpi-fun2/dinetravel/problem.tex:16]: Yes. We want to say that in the _completed_ graph with one more edge, every node should have a path to the lasagna. We're not quite sure how to rephrase this sentence more clearly without making the formalization obvious.

Comment (schmidtjonas) [reply cp23/hpi-fun2/dinetravel/problem.tex:16]: Ok, then its fine for me if you leave it like this

**Comment by schmidtjonas** [comment]
Maybe I missed it, but I think you can add more mean testcases to check all constraints. For example events causing themselves, one event causing multiple others or very small cases.

Or you could think about not giving the timeline in the input (but still asking for consistency) but only the lasagna event. A solution would be: 1. check if there are only 2 nodes with outdegree 0 and one of them is the lasagna, 2. check that there are no cycles (by either computing a topo sort yourself, or by checking if the number of components is two) 3. output the number of nodes in the lasagna component and the edge from the other root to the lasagna.
I like this problem more, but you can also leave it as it is.

**Comment by MarcelGarus** [comment]
> you could think about not giving the timeline in the input

Our goal with this problem was that it's pretty easy to solve once you correctly formalize it. So, we'd like to leave it as-is.

> Maybe I missed it, but I think you can add more mean testcases to check all constraints. For example events causing themselves, one event causing multiple others or very small cases.

Our intention with the problem statement was to exclude such cases. In particular, the following sentence guarantees all events are a tree:

> She knows all the events that she noted directly or indirectly caused vegetable lasagna to emerge.

Should we change something to make that clearer?

(Sorry for the late reply, we thought we had already sent this comment a looong time ago.)

**Comment by schmidtjonas** [comment]
Ok, I think the statement is not super easy to understand but you can leave it like it is.
Please add some form of solution notes for the presentation, where you give a formal description of the problem and a solution sketch (like our notes). Everything else looks good

**Pull request merged**